components/script/dom/bindings/constructor_hook.rs

```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::ptr;
+
+use js::gc::RootedGuard;
+use js::jsapi::{CallArgs, JSObject};
+use js::rust::HandleObject;
+
+use crate::dom::bindings::codegen::PrototypeList;
+use crate::dom::bindings::conversions::DerivedFrom;
+use crate::dom::bindings::error::throw_constructor_without_new;
+use crate::dom::bindings::htmlconstructor::call_html_constructor;
+use crate::dom::bindings::interface::get_desired_proto;
+use crate::dom::bindings::reflector::DomObject;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::utils::ProtoOrIfaceArray;
+use crate::dom::element::Element;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::window::Window;
+use crate::script_runtime::JSContext;
+
+pub unsafe fn hook_html_constructor<T: DerivedFrom<Element> + DomObject>(
+    cx: JSContext,
+    args: &CallArgs,
+    global: DomRoot<GlobalScope>,
+    proto_id: PrototypeList::ID,
+    creator: unsafe fn(JSContext, HandleObject, *mut ProtoOrIfaceArray),
+) -> bool {
+    let global = DomRoot::downcast::<Window>(global).unwrap();
+    call_html_constructor::<T>(cx, args, &global, proto_id, creator)
+}
```

Comment:
I think that this is the only call to the `call_html_constructor` so we can just change call_html_constructor instead of adding another wrapper.

Reply:
Make `call_html_constructor` accept `global: DomRoot<GlobalScope>`, then do conversion to `let global = DomRoot::downcast::<Window>(global).unwrap();` inside `call_html_constructor`.

Reply:
Given that `call_html_contructor` will simply call `html_construct`, which will again upcast it, do I pass `&GlobalScope` and do the conversion to Window inside `html_construct`?

Reply:
Yes, that would be even better.